src/Language/PureScript/TypeChecker/Types.hs

```diff
@@ -69,23 +70,38 @@ typesOf moduleName vals = do
   tys <- fmap tidyUp . liftUnifyWarnings replace $ do
     (untyped, typed, dict, untypedDict) <- typeDictionaryForBindingGroup moduleName vals
     ds1 <- parU typed $ \e -> checkTypedBindingGroupElement moduleName e dict
-    ds2 <- forM untyped $ \e -> typeForBindingGroupElement True e dict untypedDict
-    return $ ds1 ++ ds2
+    ds2 <- forM untyped $ \e -> typeForBindingGroupElement e dict untypedDict
+    return (map (\x -> (False, x)) ds1 ++ map (\x -> (True, x)) ds2)
 
-  forM tys $ \(ident, (val, ty)) -> do
+  forM tys $ \(shouldGeneralize, (ident, (val, ty))) -> do
     -- Replace type class dictionary placeholders with actual dictionaries
-    val' <- replaceTypeClassDictionaries moduleName val
+    (val', unsolved) <- replaceTypeClassDictionaries shouldGeneralize moduleName val
+    let unsolvedTypeVars = nub $ unknownsInType ty
+    -- Generalize and constrain the type
+    let generalized = generalize unsolved ty
+    -- Make sure any unsolved type constraints only use type variables which appear
+    -- unknown in the inferred type.
+    when shouldGeneralize $ do
+      tell . errorMessage $ MissingTypeDeclaration ident generalized
```

Comment:
It's been moved here, to make sure constraints get generalized in the error message.